File: leetpattern/cpp/0001_0300/0092_reverse_linked_list_ii.cc
#include <cassert>

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (!head || left == right) return head;

        ListNode dummy(0);
        dummy.next = head;
        ListNode* p0 = &dummy;

        for (int i = 0; i < left - 1; ++i) {
            p0 = p0->next;
        }

        ListNode* pre = nullptr;
        ListNode* cur = p0->next;
        int count = right - left + 1;

        while (count--) {
            ListNode* next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }

        p0->next->next = cur;
        p0->next = pre;

        return dummy.next;
    }
};

int main() {
    Solution solution;

    ListNode* head1 = new ListNode(
        1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    int left1 = 2, right1 = 4;
    ListNode* result1 = solution.reverseBetween(head1, left1, right1);
    ListNode* expected1 = new ListNode(
        1, new ListNode(4, new ListNode(3, new ListNode(2, new ListNode(5)))));
    for (ListNode *p = result1, *q = expected1; p != nullptr && q != nullptr;
         p = p->next, q = q->next) {
        assert(p->val == q->val);
    }

    ListNode* head2 = new ListNode(5);
    int left2 = 1, right2 = 1;
    ListNode* result2 = solution.reverseBetween(head2, left2, right2);
    ListNode* expected2 = new ListNode(5);
    for (ListNode *p = result2, *q = expected2; p != nullptr && q != nullptr;
         p = p->next, q = q->next) {
        assert(p->val == q->val);
    }

    return 0;
}

File: leetpattern/python/0001_0300/0092_reverse_linked_list_ii.py
"""
- Reverse a linked list from position left to position right. Return the linked list after reversing.
"""

from typing import Optional

from leetpattern.utils import ListNode, list_from_array, list_to_array


def reverseBetween(head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
    dummy = ListNode(next=head)
    p0 = dummy
    for _ in range(left - 1):
        if p0.next is None:
            break
        p0 = p0.next

    pre = None
    cur = p0.next
    for _ in range(right - left + 1):
        if not cur:
            break
        nxt = cur.next
        cur.next = pre
        pre = cur
        cur = nxt

    if p0.next:
        p0.next.next = cur
    p0.next = pre

    return dummy.next


def test_reverse_between():
    head = list_from_array([1, 2, 3, 4, 5])
    left = 2
    right = 4
    result = reverseBetween(head, left, right)
    assert list_to_array(result) == [1, 4, 3, 2, 5]

    head = list_from_array([5])
    left = 1
    right = 1
    result = reverseBetween(head, left, right)
    assert list_to_array(result) == [5]

    head = list_from_array([3, 5])
    left = 1
    right = 2
    result = reverseBetween(head, left, right)
    assert list_to_array(result) == [5, 3]
